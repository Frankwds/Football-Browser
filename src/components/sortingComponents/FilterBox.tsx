import { Button, VisuallyHidden } from "@chakra-ui/react";
import { BsSearch } from "react-icons/bs";
import "./../style.css";
import { Grid } from "@chakra-ui/react";
import { Box } from "@chakra-ui/react";
import SearchBar from "./SearchBar";
import { useState } from "react";
import { useDispatch } from "react-redux";

type Props = {
  color: string;
  accent: string;
};

/**
 * FilterBox - Component to facilitate filtering and search depending on parameters
 *
 * @param func
 * @param color
 * @param accent
 *
 * @returns {React.FC}
 *
 * Utilizing hooks to keep track of which input fields to filter on.
 * The @param func acts as a callback function to @link {../App.tsx} with the hooks.
 *
 */

const FilterBox: React.FC<Props> = ({ color, accent }) => {
  const [league, setLeague] = useState("");
  const [contry, setContry] = useState("");
  const [season, setSeason] = useState("");
  const [ht, setHt] = useState("");
  const [at, setAt] = useState("");
  const updateSearchQueries = useDispatch();

  const handleApplyFilter = (
    league: string,
    contry: string,
    season: string,
    ht: string,
    at: string
  ) => {
    // Provide given search query object
    updateSearchQueries({
      type: "UPDATE_SEARCH_DATA",
      payload: {
        league: league,
        country: contry,
        season: season,
        ht: ht,
        at: at,
      },
    });
  };

  //Renders the filterbox, with visually hidden text for improved web accessability.s
  return (
    <Box bg={color} borderRadius="lg" padding={"15px"}>
      <Grid templateRows="repeat(6, 1fr)" gap={3}>
        <Box
          bg={accent}
          fontSize="xl"
          borderRadius="lg"
          padding={"6px"}
          color={"white"}
          display="flex"
        >
          - Filter By -
        </Box>
        <SearchBar id="filterLeague" color={accent} type={"League: "} get={setLeague}>
          <VisuallyHidden>Search for a League, example: l1</VisuallyHidden>
        </SearchBar>
        <SearchBar id="searchCountry" color={accent} type={"Contry: "} get={setContry}>
          <VisuallyHidden>
            Search for a Country, example: england
          </VisuallyHidden>
        </SearchBar>
        <SearchBar id="filterSeason" color={accent} type={"Season: "} get={setSeason}>
          <VisuallyHidden>Search for a Season, example 2016</VisuallyHidden>
        </SearchBar>
        <SearchBar id="searchHT" color={accent} type={"Home Team: "} get={setHt}>
          <VisuallyHidden>
            Search for a home-team, example Stoke City
          </VisuallyHidden>
        </SearchBar>
        <SearchBar id="searchAT" color={accent} type={"Away Team: "} get={setAt}>
          <VisuallyHidden>
            Search for a away-team, example Everton
          </VisuallyHidden>
        </SearchBar>
        <Button
          id="filterAndSearchButton"
          border="2px"
          borderColor={accent}
          color={accent}
          background={"white"}
          textAlign={"center"}
          //Passing search parameters back to App.tsx
          onClick={() => handleApplyFilter(league, contry, season, ht, at)}
        >
          -Filter and search!
          <BsSearch />-
        </Button>
      </Grid>
    </Box>
  );
};

export default FilterBox;
