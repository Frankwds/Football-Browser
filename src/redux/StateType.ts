import {SearchData} from "./searchFilter/searchFilterTypes"; 

//A type interface for the only implemented query, scalable with more reducers.
export interface StateType{
    searchQueries: SearchData,
}


