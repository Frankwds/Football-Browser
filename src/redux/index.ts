import { combineReducers } from "redux";
import {searchReducer} from "./searchFilter/filterSearchReducer";


// In this case only one reducer, but very scalable.
const allReducers = combineReducers({
  searchQueries: searchReducer,
});

export default allReducers;