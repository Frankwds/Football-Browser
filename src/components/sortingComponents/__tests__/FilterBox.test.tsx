import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import FilterBox from "../FilterBox";
import renderer from "react-test-renderer";
import { createStore } from "redux";
import allReducers from "../../../redux";
import { Provider } from "react-redux";
import { ChakraProvider } from "@chakra-ui/react";

// Component input parameters
const col = "rgb(192,192,192)";
const acc = "rgb(0,128,128)";

// Test <div> container
let container: any;

// Set up redux store
const store = createStore(allReducers);

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

describe("Testing FilterBox", () => {
  it("should render without crashing", () => {
    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <ChakraProvider>
            <FilterBox color={col} accent={acc} />
          </ChakraProvider>
        </Provider>,
        container
      );
    });
  });

  it("should render with the correct fields", () => {
    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <ChakraProvider>
            <FilterBox color={col} accent={acc} />
          </ChakraProvider>
        </Provider>,
        container
      );
      expect(container.querySelector("button").textContent).toBe(
        "-Filter and search!-"
      );
    });
  });

  it("snapshot should be same as previous", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ChakraProvider>
            <FilterBox color={col} accent={acc} />
          </ChakraProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
