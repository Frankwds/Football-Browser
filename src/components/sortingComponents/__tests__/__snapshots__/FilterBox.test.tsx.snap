// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing FilterBox snapshot should be same as previous 1`] = `
Array [
  <div
    className="css-cmwzg8"
  >
    <div
      className="css-1qcq0uf"
    >
      <div
        className="css-1w6p8gg"
      >
        - Filter By -
      </div>
      <div
        className="chakra-input__group css-4302v8"
      >
        <div
          className="chakra-input__left-addon css-1en9n2p"
        >
          League: 
        </div>
        <input
          className="chakra-input css-2nurv4"
          id="filterLeague"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="All"
        />
      </div>
      <div
        className="chakra-input__group css-4302v8"
      >
        <div
          className="chakra-input__left-addon css-1en9n2p"
        >
          Contry: 
        </div>
        <input
          className="chakra-input css-2nurv4"
          id="searchCountry"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="All"
        />
      </div>
      <div
        className="chakra-input__group css-4302v8"
      >
        <div
          className="chakra-input__left-addon css-1en9n2p"
        >
          Season: 
        </div>
        <input
          className="chakra-input css-2nurv4"
          id="filterSeason"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="All"
        />
      </div>
      <div
        className="chakra-input__group css-4302v8"
      >
        <div
          className="chakra-input__left-addon css-1en9n2p"
        >
          Home Team: 
        </div>
        <input
          className="chakra-input css-2nurv4"
          id="searchHT"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="All"
        />
      </div>
      <div
        className="chakra-input__group css-4302v8"
      >
        <div
          className="chakra-input__left-addon css-1en9n2p"
        >
          Away Team: 
        </div>
        <input
          className="chakra-input css-2nurv4"
          id="searchAT"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="All"
        />
      </div>
      <button
        className="chakra-button css-hh65fl"
        id="filterAndSearchButton"
        onClick={[Function]}
        type="button"
      >
        -Filter and search!
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
          />
        </svg>
        -
      </button>
    </div>
  </div>,
  <span />,
]
`;
